public_html/login.php

```diff
@@ -29,6 +30,135 @@
+
+
+$gUserAgent = @$CONFIG['oauth']['agent'];
+$gConsumerKey = @$CONFIG['oauth']['consumerKey'];
+$gConsumerSecret = @$CONFIG['oauth']['consumerSecret'];
+
+// Load the user token (request or access) from the session
+$gTokenKey = '';
+$gTokenSecret = '';
+session_name('UTRSLogin');
+session_start();
+
+if ( isset( $_GET['oauth_verifier'] ) && $_GET['oauth_verifier'] ) {
+    $gTokenKey = $_SESSION['tokenKey'];
+    $gTokenSecret = $_SESSION['tokenSecret'];
+    fetchAccessToken();
+    $payload = doIdentify();
+
+    $is_admin = in_array("sysop", $payload->groups);
+    $is_check = in_array("checkuser", $payload->groups);
+    $username = $payload->username;
+
+    if ($is_admin === TRUE && $payload->confirmed_email === TRUE) {
+        session_name('UTRSLogin');
+        session_start();
+        $_SESSION['user'] = $username;
+        $_SESSION['oauth'] = TRUE;
+
+        $db = connectToDB(true);
+        $query = $db->prepare('
+                SELECT userID FROM user
+                WHERE username = :username');
+
+        $result = $query->execute(array(
+                ':username'	=> $username));
+
+        if($result === false){
+                $error = var_export($query->errorInfo(), true);
+                debug('ERROR: ' . $error . '<br/>');
+                throw new UTRSDatabaseException($error);
+        }
+        $data = $query->fetch(PDO::FETCH_ASSOC);
+        $query->closeCursor();
+        if ($data['userID'] === NULL) {
+            $user = new User(array(
+                'wikiAccount' => $username,
+                'diff' => "",
+                'username' => $username,
+                'email' => $payload->email,
+            ), false, array(
+                'checkuser' => $is_check,
+            ));
+            debug('object created<br/>');
+        } else {
+            $user = User::getUserById($data['userID']);
+            if ($user->isCheckuser() !== $is_check || $user->getEmail() !== $payload->email) {
+                // XXX: Logging?
+                $query = $db->prepare("
+                        UPDATE user
+                        SET checkuser = :checkuser,
+                            email = :email
+                        WHERE userID = :userID");
+                $result = $query->execute(array(
+                        ':checkuser' => (bool)$is_check,
+                        ':email' => $payload->email,
+                        ':userID' => (int)$data['userID']));
+                if(!$result){
+                    $error = var_export($query->errorInfo(), true);
+                    debug('ERROR: ' . $error . '<br/>');
+                    throw new UTRSDatabaseException($error);
+                }
+            }
+        }
+        header("Location: " . "home.php");
+        exit;
+    } else if ($is_admin === TRUE) {
+        $errors = 'You need to have a confirmed email address set in MediaWiki to use UTRS.';
+    } else {
+        $errors = 'Only administrators can review requests on UTRS.';

```

Comment:
Watch this case happen, someone needs access for security/emergency reasons from the WMF (or hell uses their WMF account to signup) and doesn't have sysop privileges. How is this remedied?